api/scenario_api_simulator/src/scenario_api_simulator.cpp

```diff
@@ -518,56 +539,49 @@ void ScenarioAPISimulator::updateNPC()
   inputNPCStopState(npc_stop_states);
 }
 
-bool ScenarioAPISimulator::getNPC(const std::string & name, npc_simulator::Object & obj)
+bool ScenarioAPISimulator::getNPC(const std::string & name, npc_simulator::msg::Object & obj)
 {
-  if (!checkValidNPC(name))
-  {
-    ROS_WARN_STREAM("Invalid NPC name '" << name << "' requested.");
+  if (!checkValidNPC(name)) {
+    RCLCPP_WARN_STREAM(get_logger(), "Invalid NPC name '" << name << "' requested.");
     return false;
-  }
-  else
-  {
-    npc_simulator::GetObject srv;
-
-    srv.request.object_id = uuid_map_[name];
+  } else {
+    auto req = std::make_shared<npc_simulator::srv::GetObject::Request>();
+    req->object_id = uuid_map_.at(name);
 
-    if (not client_.call(srv) or not srv.response.success)
-    {
-      ROS_WARN_STREAM("Failed to get NPC");
-      return false;
-    }
-    else
-    {
-      obj = srv.response.object;
+    auto res_future = client_->async_send_request(req);
```

Comment:
This is the first node with service calls besides `pose_initializer`, and already there, these service calls were somewhat [problematic (link)](https://github.com/tier4/Pilot.Auto/pull/11#discussion_r502286589). There is no good synchronous API, similar to `waitForTransform()` – it might be that `wait_for()` works again, like with `waitForTransform()`, otherwise we're stuck with callbacks, but I haven't tried it yet. 

One alternative is to replace the client with an actual instance of NPCSimulatorNode and call its service handler (`getObject()`) directly. That would mean having nested nodes, which would need to be added to the executor manually, but we have nested nodes anyway, for instance in [the next package, scenario_api](https://github.com/tier4/scenario_runner.iv.universe/blob/d6ec250b2ff9a0ae879ca19a27a7fc3599af4a0a/api/scenario_api/include/scenario_api/scenario_api_core.h#L305-L307).

Reply:
Okay, then let's create actual instance of NPCSimulatorNode for this one.
I will leave it to you if you implement as a nested node or make changes to npc_simulator to take node pointer as input to not register it as new node.

Reply:
Thanks @mitsudome-r – I checked and `wait_for()` always times out, even with a multithreaded executor.

Reply:
To recap, having a nested node means we'd have something like this:
* `std::shared_ptr<NPCSimulatorNode> npc_simulator_;` member inside `ScenarioAPISimulator`
* `add_sub_node(rclcpp::Executor& exec)` (and corresponding `remove_sub_node`) function inside `ScenarioAPISimulator` that adds this node to the executor
* `add_sub_node(rclcpp::Executor& exec)` (and corresponding `remove_sub_node`) function inside `ScenarioAPI` that
calls `simulator_api_->add_sub_node(exec)`
* A call to this function by `main()`

On the other hand, we could also make `npc_simulator` an "extension node" like `ScenarioAPISimulator`, `ScenarioAPIAutoware` etc., i.e. pass in a node in the constructor instead of inheriting from `rclcpp::Node`. Then passing around the executor wouldn't be necessary. Would that be preferable?

Reply:
fyi @fred-apex-ai

Reply:
I implemented the second option here (a few TODOs remaining): https://github.com/tier4/planning_simulator.iv.universe/pull/8